python/modules/IcePy/Communicator.cpp

```diff
@@ -1606,6 +1594,22 @@ IcePy::getCommunicatorWrapper(const Ice::CommunicatorPtr& communicator)
     }
 }
 
+Ice::SliceLoaderPtr
+IcePy::getSliceLoader(const Ice::CommunicatorPtr& communicator)
+{
+    // It's in the map until the communicator is destroyed.
+    auto p = sliceLoaderMap.find(communicator);
+    assert(p != sliceLoaderMap.end());
+    return p->second;
+}
+
+void
+IcePy::removeSliceLoader(const Ice::CommunicatorPtr& communicator)
+{
+    // Called by destroy[Async], which can be called multiple times.
+    sliceLoaderMap.erase(communicator);
```

Comment:
This is only safe if the deallocation of the slice loader doesn't release the GIL. It is a bit unpredictable when the GIL is released and reacquired. I think it would be clear to use a regular C++ lock for sliceLoaderMap, specially because the Slice loader can be provided by the user.

Reply:
Sorry, I'm not quite following what you mean here.

Reply:
The PythonSlice loader can hold a user object, when this object is released it can do:

```
// Pseudo code
{
  AllowThreads allow; // <- Now another thread can reenter removeSliceLoader
  ....
}
// Continue with the GIL locked
```

Reply:
Yes, the Python SliceLoader, when deallocated, can temporarily release the GIL.

I don't see why this is a concern here. I believe all this code, part of the "Python communicator destroy", always run prior to the finalization of the Python interpreter.

Reply:
> I don't see why this is a concern here

My concern is that `sliceLoaderMap.erase` is only protected by the GIL, if it can be released while erase deallocates the SliceLoader, another thread can reenter erase.